### PR TITLE
[Merged by Bors] - feat: binomial theorem for subtraction

### PR DESCRIPTION
**Pull request by ahhwuhu**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/LegendrePoly.lean [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+open scoped Nat
+open BigOperators
+
+variable {R : Type*}
+
+namespace Polynomial
+
+noncomputable abbrev Legendre (n : ℕ) : ℝ[X] :=
+  C (1 / n ! : ℝ) * derivative^[n] (X ^ n * (1 - X) ^ n)
```
Comment (YaelDillies) [inline Mathlib/Analysis/LegendrePoly.lean:33]:
```suggestion
  C n!⁻¹ * derivative^[n] (X ^ n * (1 - X) ^ n)
```

File: Mathlib/Analysis/LegendrePoly.lean [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+lemma deriv_one_sub_X {n i : ℕ} : (⇑derivative)^[i] ((1 - X) ^ n : ℝ[X]) =
+    (-1) ^ i * (n.descFactorial i) • ((1 - X) ^ (n - i)) := by
+  rw [show (1 - X : ℝ[X]) ^ n = (X ^ n : ℝ[X]).comp (1 - X) by simp,
+    Polynomial.iterate_derivative_comp_one_sub_X (p := X ^ n),
+    Polynomial.iterate_derivative_X_pow_eq_smul]
+  rw [Algebra.smul_def, algebraMap_eq, map_natCast]
+  simp
+
+lemma legendre_eval_symm {n : ℕ} {x : ℝ} : eval x (Legendre n) =
+    (-1) ^ n * eval (1 - x) (Legendre n) := by
```
Comment (YaelDillies) [inline Mathlib/Analysis/LegendrePoly.lean:117]:
```suggestion
lemma legendre_eval_symm (n : ℕ) (x : ℝ) :
    eval x (Legendre n) = (-1) ^ n * eval (1 - x) (Legendre n) := by
```

File: Mathlib/Data/Nat/Choose/Sum.lean [hunk 3]
```diff
@@ -74,6 +75,18 @@ theorem add_pow [CommSemiring R] (x y : R) (n : ℕ) :
     (x + y) ^ n = ∑ m ∈ range (n + 1), x ^ m * y ^ (n - m) * n.choose m :=
   (Commute.all x y).add_pow n
 
+/-- An special case of the **binomial theorem** -/
+theorem sub_pow [CommRing R] (x : R) (y : R) (n : ℕ) :
+    (x - y) ^ n = ∑ m ∈ range (n + 1), (n.choose m) • x ^ m * (- 1) ^ (n - m) * y ^ (n - m) := by
```
Comment (YaelDillies) [inline Mathlib/Data/Nat/Choose/Sum.lean:80]:
```suggestion
    (x - y) ^ n = ∑ m ∈ range (n + 1), n.choose m • x ^ m * (- 1) ^ (n - m) * y ^ (n - m) := by
```

File: Mathlib/Data/Nat/Choose/Sum.lean [hunk 4]
```diff
@@ -74,6 +75,18 @@ theorem add_pow [CommSemiring R] (x y : R) (n : ℕ) :
...
+  rw [← Mathlib.Tactic.RingNF.add_neg, add_pow]
```
Comment (YaelDillies) [inline Mathlib/Data/Nat/Choose/Sum.lean:81]: Please don't use lemmas from tactic implementations
```suggestion
  rw [sub_eq_add_neg, add_pow]
```

File: Mathlib/Data/Nat/Choose/Sum.lean
Comment (YaelDillies) [inline Mathlib/Data/Nat/Choose/Sum.lean:1]: Can you make this be a separate PR?

Comment (ahhwuhu) [reply Mathlib/Data/Nat/Choose/Sum.lean:1]: okk

File: Mathlib/Data/Nat/Choose/Sum.lean [hunk 5]
```diff
@@ -74,6 +75,27 @@ theorem add_pow [CommSemiring R] (x y : R) (n : ℕ) :
     (x + y) ^ n = ∑ m ∈ range (n + 1), x ^ m * y ^ (n - m) * n.choose m :=
   (Commute.all x y).add_pow n
 
+/-- A special case of the **binomial theorem** -/
+theorem sub_pow [CommRing R] (x : R) (y : R) (n : ℕ) :
+    (x - y) ^ n = ∑ m ∈ range (n + 1), n.choose m • x ^ m * (- 1) ^ (n - m) * y ^ (n - m) := by
```
Comment (YaelDillies) [inline Mathlib/Data/Nat/Choose/Sum.lean:80]:
```suggestion
    (x - y) ^ n = ∑ m ∈ range (n + 1), (-1) ^ (m + n) * n.choose m • x ^ m * y ^ (n - m) := by
```
might be more canonical?

Comment (ahhwuhu) [reply Mathlib/Data/Nat/Choose/Sum.lean]: ok! I get it!

File: Mathlib/Data/Nat/Choose/Sum.lean [hunk 6]
```diff
@@ -74,6 +75,32 @@ theorem add_pow [CommSemiring R] (x y : R) (n : ℕ) :
...
+  ring_nf
+  nth_rewrite 3 [mul_assoc]
+  congr
+  rw [← mul_one (a := (-1) ^ (n - m))]
+  nth_rewrite 2 [show (1 : R) = (-1) ^ (2 * m) by simp]
+  rw [← pow_add, ← pow_add, show n - m + 2 * m = m + n by omega]
+
+/-- A special case of the sub_pow -/
+theorem sub_square_pow [CommRing R] (x : R) (n : ℕ) :
+    (x - x ^ 2) ^ n = ∑ m ∈ range (n + 1), (- 1) ^ m * n.choose m • x ^ (n + m) := by
```
Comment (YaelDillies) [inline Mathlib/Data/Nat/Choose/Sum.lean:95]: I don't really see the value in this lemma. You get this just by applying `sub_pow` and rewriting the terms in the obvious fashion. Can you inline it in your application?

Comment (ahhwuhu) [reply Mathlib/Data/Nat/Choose/Sum.lean:95]: emmmmm. I get it! It must be ok!

Comment (ahhwuhu) [reply Mathlib/Data/Nat/Choose/Sum.lean:95]: I worried that my proof is too long, so I separate it. ok. I'll inline it in my lemma.

**Comment by ahhwuhu** [comment]
ok! I fix all problem! What should I do next?

File: Mathlib/Data/Nat/Choose/Sum.lean [hunk 7]
```diff
@@ -74,6 +75,21 @@ theorem add_pow [CommSemiring R] (x y : R) (n : ℕ) :
...
+  apply Finset.sum_congr rfl
+  intro m hm
+  simp only [mem_range] at hm
+  rw [neg_pow]
+  ring_nf
+  nth_rewrite 3 [mul_assoc]
+  congr
+  rw [← mul_one (a := (-1) ^ (n - m))]
+  nth_rewrite 2 [show (1 : R) = (-1) ^ (2 * m) by simp]
+  rw [← pow_add, ← pow_add, show n - m + 2 * m = m + n by omega]
```
Comment (YaelDillies) [inline Mathlib/Data/Nat/Choose/Sum.lean:91]:
```suggestion
  congr! 1 with m hm
  have : (-1 : R) ^ (n - m) = (-1) ^ (n + m) := by
    rw [mem_range] at hm
    simp [show n + m = n - m + 2 * m by omega, pow_add]
  rw [neg_pow, this]
  ring
```

Comment (ahhwuhu) [reply Mathlib/Data/Nat/Choose/Sum.lean]: thanks!

**Comment by ahhwuhu** [comment]
Thank you for your suggestion. I learn a lot. Then how can I merge this PR?

**Review by YaelDillies — APPROVED** [review]
Neither of us can merge this PR. We need the approval of a maintainer. Luckily, I have mystical powers that let me invoke one solely by ushering the words

maintainer merge

**Comment by ahhwuhu** [comment]
> Neither of us can merge this PR. We need the approval of a maintainer. Luckily, I have mystical powers that let me invoke one solely by ushering the words
> 
> maintainer merge

hahahahahaha! That's so cool! I'm so lucky to meet you to review my PR and I learn so much from you! Thank you very much Combinatorics math genius!!!

File: Mathlib/Data/Nat/Choose/Sum.lean [hunk 8]
```diff
@@ -74,6 +75,17 @@ theorem add_pow [CommSemiring R] (x y : R) (n : ℕ) :
     (x + y) ^ n = ∑ m ∈ range (n + 1), x ^ m * y ^ (n - m) * n.choose m :=
   (Commute.all x y).add_pow n
 
+/-- A special case of the **binomial theorem** -/
+theorem sub_pow [CommRing R] (x : R) (y : R) (n : ℕ) :
```
Comment (b-mehta) [inline Mathlib/Data/Nat/Choose/Sum.lean:79]:
```suggestion
theorem sub_pow [CommRing R] (x y : R) (n : ℕ) :
```

File: Mathlib/Data/Nat/Choose/Sum.lean [hunk 9]
```diff
@@ -74,6 +75,17 @@ theorem add_pow [CommSemiring R] (x y : R) (n : ℕ) :
...
+    (x - y) ^ n = ∑ m ∈ range (n + 1), (-1) ^ (m + n) * n.choose m • x ^ m * y ^ (n - m) := by
```
Comment (b-mehta) [inline Mathlib/Data/Nat/Choose/Sum.lean:80]:
```suggestion
    (x - y) ^ n = ∑ m ∈ range (n + 1), (-1) ^ (m + n) * x ^ m * y ^ (n - m) * n.choose m  := by
```
for consistency with the previous version

File: Mathlib/Data/Nat/Choose/Sum.lean [hunk 10]
```diff
@@ -10,6 +10,7 @@ import Mathlib.Algebra.Order.BigOperators.Group.Finset
 import Mathlib.Data.Nat.Choose.Basic
 import Mathlib.Tactic.Linarith
 import Mathlib.Tactic.Ring
+import Mathlib.Tactic.FieldSimp
```
Comment (b-mehta) [inline Mathlib/Data/Nat/Choose/Sum.lean:13]: Do you still need this import?

**Review by b-mehta — APPROVED** [review]

File: Mathlib/Data/Nat/Choose/Sum.lean (hunk 10 above)
Comment (ahhwuhu) [reply Mathlib/Data/Nat/Choose/Sum.lean:13]: oh! I don't need.

**Comment by b-mehta** [comment]
bors merge

**Pull request closed without merging**